cloudevents/working-drafts/minor-versions.md

```diff
@@ -0,0 +1,337 @@
+```
+
+### "Changing" event versions
+
+An SDK may provide the ability to change the specversion of an event,
+or (preferred) create a new object based on the old one. When "upgrading" an event's specversion,
+the SDK should perform validation as if parsing the event from scratch. Examples:
+
+```csharp
+// Successful upgrade: no xyz value
+var evt10 = new CloudEvent(CloudEventsSpecVersion.V1_0) { Id = "..." };
+var evt11 = evt10.WithSpecVersion(CloudEventsSpecVersion.V1_1);
+
+// Successful upgrade: valid value for xyz
+var evt10 = new CloudEvent(CloudEventsSpecVersion.V1_0)
+{
+    Id = "...",
+    ["xyz"] = 10
+};
+evt11 = evt10.WithSpecVersion(CloudEventsSpecVersion.V1_1);
+
+// Failed upgrade: invalid value for xyz
+var evt10 = new CloudEvent(CloudEventsSpecVersion.V1_0)
+{
+    Id = "...",
+    ["xyz"] = "test"
+};
+evt11 = evt10.WithSpecVersion(CloudEventsSpecVersion.V1_1);
+
+// Successful downgrade
+var evt11 = new CloudEvent(CloudEventsSpecVersion.V1_1)
+{
+    Id = "...",
+    Xyz = 10
+};
+evt10 = evt11.WithSpecVersion(CloudEventsSpecVersion.V1_0);
+// Access to evt10.Xyz would fail as described above;
+// access to evt10["xyz"] would return 10.
+// Setting evt10["xyz"] to "test" or 123 could fail or succeed - it's
+// up to the SDK to decide (and document!) whether downgrading retains constraints
+// for attributes that were present in the original event.
+// TODO: What about if Xyz hadn't been present before? Should evt10["xyz"] = "test" fail then?
+```
+
+Until we have a new major version, we probably don't want to specify whether
+SDKs should allow upgrade/downgrade across major versions.
+
+### Handling unknown versions
+
+An SDK which supports minor versions **MUST** be able to parse events with
+a minor version of which it's unaware, but **MUST NOT** parse events with a
+*major* version of which it's unaware. In other words, a v1.1-aware SDK must
+be able to parse a v1.2 event, but must reject any attempt to parse a v2.0 event.
+This is because a v2.0 event might use an entirely different format or type system,
+whereas a v1.2 event can always be understood in the v1.x type system, and all
+attributes should be valid in the context of v1.1 and v1.0.
+
+All SDKs must support all minor versions for each major version they support, up
+to the highest minor version they're aware of. (So when spec version v1.2 is published, it's fine for
+an SDK to only support v1.0 and v1.1, but it must not support *only* v1.1 or *only* v1.2.)
+
+When parsing an event with an unknown minor version:
+
+- All attribute values must be retained, with as much type information as is available
+  (defaulting to string when type information is not available)
+- Values of unknown attributes must be readable as if they're regular extension attributes
+  (so a minor-version aware SDK that only supports v1.0 must allow access to `xyz` of
+  a v1.1 event after receiving it)
+- All attribute values must be validated according to the highest-known minor version
```

Comment:
All attribute values must be validated according to the CE specversion specified in the incoming CE, or if that minor version is unknown to the SDK then it must use the highest-known minor version's semantics.

Reply:
I think we're in agreement here, and it's just a matter of being clearer in the wording - is that right?